pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -211,21 +211,18 @@ func (r *LazyBinaryReader) LabelNames() ([]string, error) {
 }
 
 // EagerLoadIndexHeadersSnapshot eagerly load index headers from the stored lazyLoadedHeadersSnapshot
-func (r *LazyBinaryReader) EagerLoadIndexHeadersSnapshot(lazyLoadedHeadersSnapshot *lazyLoadedHeadersSnapshot) {
-	if lazyLoadedHeadersSnapshot != nil {
-		usedAtSnapshot := lazyLoadedHeadersSnapshot.IndexHeaderLastUsedTime[r.blockID]
-		if usedAtSnapshot > 0 {
-			r.readerMx.RLock()
-			defer r.readerMx.RUnlock()
-
-			if err := r.load(); err != nil {
-				level.Warn(r.logger).Log("msg", "eager loading of lazy loaded index-header failed; skipping", "err", err)
-				return
-			}
-
-			r.usedAt.Store(usedAtSnapshot)
-			r.metrics.eagerLoadCount.Inc()
+func (r *LazyBinaryReader) EagerLoadIndexHeadersSnapshot(usedAtSnapshot int64) {
+	if usedAtSnapshot > 0 {
+		r.readerMx.RLock()
+		defer r.readerMx.RUnlock()
+
+		if err := r.load(); err != nil {
+			level.Warn(r.logger).Log("msg", "eager loading of lazy loaded index-header failed; skipping", "err", err)
+			return
 		}
+
+		r.usedAt.Store(usedAtSnapshot)
```

Comment:
I'm trying to think through the implications of this (using the "used at" time from the previous process) - I'm wondering if we should reset it to the current time.

I'm imagining scenarios like this:

* If eagerly loading all index headers takes longer than the idle timeout, then we'll start unloading headers before we start accepting traffic (Is this a realistic scenario though? The default timeout is 60m)
* If the process was offline for longer than the idle timeout, depending on the traffic pattern and how long it takes to eagerly load the index headers, we may end up unloading  index headers before we have a chance to use them. (We'll do a first sweep of idle readers after `idle timeout / 10`, so with the default config, we'll do a sweep 6m after starting.)

Reply:
It is really great point. Actually somehow I couldn't recall why I am having usedAt in the lazyLoaded snapshot. It seems we better just set the usedAt to the current time every time.

@dimitarvdimitrov do you remember or do you think is there any reason for using usedAt time from the previous process? Should I just reset usedAt to time.Now during eager loading?

Reply:
I decided to set it to time.Now. I can change it again after further discussion if needed.

Reply:
this will change with https://github.com/grafana/mimir/pull/5608/files. After that gets merged I think we can get away with just calling `getOrLoadReader` in `EagerLoad` (or removing EagerLoad altogether since it's just an "alias"?)